File: painterly/density.py
"""密度图：按画面"密度"调度笔刷——轮廓和狭窄处小硬刷，开阔低反差处大软刷。

密度 = 归一化( 模糊(梯度幅值) + 模糊(Posterize 边界) )，
密度高 → 笔刷小、用硬刷；密度低 → 笔刷大、用软刷。
"""
import numpy as np
import cv2


def density_map(gx: np.ndarray, gy: np.ndarray, poster_edge: np.ndarray,
                sigma: float = 8.0, edge_weight: float = 1.0) -> np.ndarray:
    mag = np.hypot(gx, gy)
    mag = mag / (np.percentile(mag, 98) + 1e-8)  # 按分位数归一，避免个别极值把整体压扁
    d = cv2.GaussianBlur(mag, (0, 0), sigma) + edge_weight * cv2.GaussianBlur(poster_edge, (0, 0), sigma)
    d = d / (np.percentile(d, 95) + 1e-8)
    return np.clip(d, 0, 1).astype(np.float32)


def density_vis(density: np.ndarray, hard_threshold: float) -> np.ndarray:
    """红 = 高密度/小硬刷区，蓝 = 低密度/大软刷区。"""
    h, w = density.shape
    vis = np.zeros((h, w, 3), np.float32)
    hard = density >= hard_threshold
    vis[..., 0] = np.where(hard, density, 0)          # R
    vis[..., 2] = np.where(~hard, 1.0 - density, 0)   # B
    return (vis * 255).astype(np.uint8)


def brush_size_at(density: np.ndarray, size_min: float, size_max: float,
                  gamma: float = 0.7) -> np.ndarray:
    """密度 → 笔刷半径的映射（密度高 → 小）。gamma<1 让中等密度也偏向小刷子，
    轮廓才收得住。"""
    return size_max + (size_min - size_max) * np.power(density, gamma)

File: crates/painterly-core/src/density.rs
//! 密度マップ: 画面の「描き込み密度」に応じてブラシを持ち替える——
//! 輪郭や狭い場所は小さなハードブラシ、広く低コントラストな場所は大きなソフトブラシ。
//!
//! 密度 = 正規化( ぼかし(勾配強度) + ぼかし(ポスタライズ境界) )
//! Python 版 painterly/density.py の移植。

use crate::buf::{gaussian_blur, quantile, Gray};

pub fn density_map(gx: &Gray, gy: &Gray, poster_edge: &Gray, sigma: f32, edge_weight: f32) -> Gray {
    let (w, h) = (gx.w, gx.h);
    let mut mag = Gray::new(w, h);
    for i in 0..w * h {
        mag.data[i] = gx.data[i].hypot(gy.data[i]);
    }
    // 分位数で正規化して極値に全体が潰されるのを防ぐ
    let p98 = quantile(&mag.data, 0.98) + 1e-8;
    for v in &mut mag.data {
        *v /= p98;
    }
    let b1 = gaussian_blur(&mag, sigma);
    let b2 = gaussian_blur(poster_edge, sigma);
    let mut d = Gray::new(w, h);
    for i in 0..w * h {
        d.data[i] = b1.data[i] + edge_weight * b2.data[i];
    }
    let p95 = quantile(&d.data, 0.95) + 1e-8;
    for v in &mut d.data {
        *v = (*v / p95).clamp(0.0, 1.0);
    }
    d
}

/// 赤 = 高密度（小さいハードブラシ）、青 = 低密度（大きいソフトブラシ）
pub fn density_vis(density: &Gray, hard_threshold: f32) -> image::RgbImage {
    let mut out = image::RgbImage::new(density.w as u32, density.h as u32);
    for (i, p) in out.pixels_mut().enumerate() {
        let d = density.data[i];
        if d >= hard_threshold {
            p.0 = [(d * 255.0) as u8, 0, 0];
        } else {
            p.0 = [0, 0, ((1.0 - d) * 255.0) as u8];
        }
    }
    out
}

/// 密度 → ブラシ半径（高密度ほど小さく）。gamma < 1 で中密度も小さめに寄せ、
/// 輪郭の締まりを保つ
pub fn brush_size_at(density: f32, size_min: f32, size_max: f32) -> f32 {
    const GAMMA: f32 = 0.7;
    size_max + (size_min - size_max) * density.powf(GAMMA)
}
